zeppelin-web/src/app/notebook/paragraph/paragraph.controller.js

```diff
@@ -436,15 +445,24 @@ angular.module('zeppelinWebApp')
         // not applying emacs key binding while the binding override Ctrl-v. default behavior of paste text on windows.
       }
 
-      var sqlModeTest = /^%(\w*\.)?\wql/;
+      $scope.setParagraphMode = function(session, pos) {
+          var paragraphText = session.getValue();
 
-      $scope.setParagraphMode = function(session, paragraphText) {
-    	  if (sqlModeTest.test(String(paragraphText))) {
-        	  session.setMode(editorMode.sql);
-          } else if ( String(paragraphText).startsWith('%md')) {
-        	  session.setMode(editorMode.markdown);
-          } else {
-        	  session.setMode(editorMode.scala);
+          // Evaluate the mode only if the change is at the beging of the paragraph
+          if (pos.row === 0 && pos.column < 30) {
+             var newMode = editorModes[0][1];
+
+             for (var i = 0, len = editorModes.length; i < len; i++ ) {
```

Comment:
it is possible to set a default interpreter and not specific %something entirely.
perhaps check for the first non-whitespace character and that it is % first?